pkg/controller/velerobackup/velerobackup_controller_test.go

```diff
@@ -139,14 +161,219 @@ func TestReconcile(t *testing.T) {
+
+	tests := []struct {
+		name                 string
+		getHashOfObjectsFunc controllerutils.ChecksumOfObjectsFunc
+		existingObjects      []runtime.Object
+		expectedHiveObjects  []*hiveObject
+		expectedError        error
+	}{
+		{
+			name:                "No Hive objects",
+			existingObjects:     []runtime.Object{},
+			expectedHiveObjects: []*hiveObject{},
+			expectedError:       nil,
+		},
+		{
+			name: "No Changed Hive objects",
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+				testsyncset.Build(unchangedSyncSetBase()),
+			},
+			expectedHiveObjects: []*hiveObject{},
+			expectedError:       nil,
+		},
+		{
+			name: "1 Changed Hive object",
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+				testclusterdeployment.Build(changedClusterDeploymentBase()),
+			},
+			expectedHiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			},
+			expectedError: nil,
+		},
+		{
+			name: "Multiple Changed Hive objects",
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(changedClusterDeploymentBase()),
+				testclusterdeployment.Build(unchangedClusterDeploymentBase()),
+				testsyncset.Build(changedSyncSetBase()),
+				testsyncset.Build(unchangedSyncSetBase()),
+			},
+			expectedHiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+				ro2ho(testsyncset.Build(changedSyncSetBase())),
+			},
+			expectedError: nil,
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			// Arrange
+			r := fakeClientReconcileBackup(test.existingObjects)
+
+			// Act
+			actualHiveObjects, actualError := r.getModifiedHiveObjectsInNamespace(namespace)
+
+			// Assert
+			assert.Equal(t, len(test.expectedHiveObjects), len(actualHiveObjects))
+			assertHiveObjectsEqual(t, test.expectedHiveObjects, actualHiveObjects)
+			assert.Equal(t, test.expectedError, actualError)
+		})
+	}
+}
+
+func TestUpdateHiveObjectsLastBackupChecksum(t *testing.T) {
+	apis.AddToScheme(scheme.Scheme)
+
+	tests := []struct {
+		name            string
+		hiveObjects     []*hiveObject
+		existingObjects []runtime.Object
+		expectedObjects []runtime.Object
+		expectedError   error
+	}{
+		{
+			name:            "No HiveObjects",
+			hiveObjects:     []*hiveObject{},
+			existingObjects: []runtime.Object{},
+			expectedObjects: []runtime.Object{},
+			expectedError:   nil,
+		},
+		{
+			name: "1 changed ClusterDeployment",
+			hiveObjects: []*hiveObject{
+				ro2ho(testclusterdeployment.Build(changedClusterDeploymentBase())),
+			},
+			existingObjects: []runtime.Object{
+				testclusterdeployment.Build(changedClusterDeploymentBase()),
+			},
+			expectedObjects: []runtime.Object{
+				testclusterdeployment.Build(changedClusterDeploymentBase(), testclusterdeployment.Generic(generic.WithBackupChecksum(defaultChecksumFunc))),
```

Comment:
I like this much better than what you had before. I would still prefer to see tests that were more explicit about what they were testing and less dependent upon the behavior of other components (namely, be explicit that I expect this particular checksum to be set, and do not rely on things like `newHiveObject` and `WithBackupChecksum` to determine the checksum behind the scenes). I will not hold up this PR for that wish, though.

Reply:
I get your point, I lean more on the "make utils that make creating tests easy" side. Maybe this is a philosophical discussion that we should have after standup sometime.